File: leetcode/15-3sum.py
# https://leetcode.com/problems/3sum/ (top-interview-questions)

from typing import List


class Solution:
	def threeSum(self, nums: List[int]) -> List[List[int]]:
		nums.sort()
		triplets = list()

		for i in range(len(nums)-2):
			# the solution set must not contain duplicate triplets.
			if i > 0 and nums[i] == nums[i-1]:
				continue
			# find any two complement...
			l, r = i+1, len(nums)-1
			while l < r:
				triplet_sum = nums[i] + nums[l] + nums[r]
				if triplet_sum < 0:
					l += 1
				elif triplet_sum > 0:
					r -= 1
				else:
					triplets.append([nums[i], nums[l], nums[r]])
					# the solution set must not contain duplicate triplets.
					l += 1
					while nums[l] == nums[l-1] and l < r:
						l += 1
					r -= 1
					while nums[r] == nums[r+1] and l < r: 
						r -= 1

		return triplets

File: leetcode/15-3sum.js
// https://leetcode.com/problems/3sum/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  nums.sort((a, b) => (a - b));
  let triplets = [];
  for (let i = 0; i < nums.length-2; i++) {
    if (i > 0 && nums[i-1] === nums[i]) {
      continue;
    }

    let l = i+1;
    let r = nums.length - 1;

    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r];
      if (sum === 0) {
        triplets.push([nums[i], nums[l], nums[r]]);
        l += 1;
        while (nums[l-1] === nums[l] && l < r) l += 1;
        r -= 1;
        while (nums[r] === nums[r+1] && l < r) r -= 1;
      } else if (sum < 0) {
        l += 1;
      } else {
        r -= 1;
      }
    }

  }

  return triplets;

};


threeSum([-1,0,1,2,-1,-4]);